lib/dockerregistry/inventory.go

```diff
@@ -1531,6 +1572,152 @@ func (sc *SyncContext) GarbageCollect(
+
+		var populateRequests PopulateRequests = func(
+			sc *SyncContext,
+			reqs chan<- stream.ExternalRequest,
+			wg *sync.WaitGroup) {
+
+			for _, registry := range mfest.Registries {
+				// Skip over any registry that does not match the regName we want to
+				// wipe.
+				if registry.Name != regName {
+					continue
+				}
+				for imageName, digestTags := range sc.Inv[registry.Name] {
+					for digest, _ := range digestTags {
+						mediaType, ok := sc.DigestMediaType[digest]
+						if !ok {
+							fmt.Println("could not detect MediaType of digest", digest)
+							continue
+						}
+						if !predicate(mediaType) {
+							fmt.Printf("skipping digest %s mediaType %s\n", digest, mediaType)
+							continue
+						}
+						var req stream.ExternalRequest
+						req.StreamProducer = mkProducer(
+							registry,
+							imageName,
+							digest)
+						req.RequestParams = PromotionRequest{
+							Delete,
+							"",
+							registry.Name,
+							registry.ServiceAccount,
+
+							// No source image name, because tag deletions
+							// should only delete the what's in the
+							// destination registry
+							ImageName(""),
+
+							imageName,
+							digest,
+							"",
+							"",
+						}
+						wg.Add(1)
+						reqs <- req
+					}
+				}
+			}
+		}
+		return populateRequests
+	}
+
+	var processRequest ProcessRequest
+	var processRequestReal ProcessRequest = func(
+		sc *SyncContext,
+		reqs chan stream.ExternalRequest,
+		requestResults chan<- RequestResult,
+		wg *sync.WaitGroup,
+		mutex *sync.Mutex) {
+
+		for req := range reqs {
+			reqRes := RequestResult{Context: req}
+			jsons, errors := getJSONSFromProcess(req)
+			if len(errors) > 0 {
+				reqRes.Errors = errors
+				requestResults <- reqRes
+				// Don't skip over this request, because stderr here is ignorable.
+				// continue
+			}
+			for _, json := range jsons {
+				sc.Info("DELETED image:", json)
+			}
+			reqRes.Errors = errors
+			requestResults <- reqRes
+			wg.Add(-1)
+		}
+	}
+
+	captured := make(CapturedRequests)
+
+	if sc.DryRun {
+		processRequestDryRun := MkRequestCapturer(&captured)
+		processRequest = processRequestDryRun
+	} else {
+		processRequest = processRequestReal
+	}
+
+	if customProcessRequest != nil {
+		processRequest = *customProcessRequest
+	}
+
+	var isEqualTo (func(cr.MediaType) func(cr.MediaType) bool) = func(want cr.MediaType) func(cr.MediaType) bool {
+		return func(got cr.MediaType) bool {
+			return want == got
+		}
+	}
+
+	var isNotEqualTo (func(cr.MediaType) func(cr.MediaType) bool) = func(want cr.MediaType) func(cr.MediaType) bool {
+		return func(got cr.MediaType) bool {
+			return want != got
+		}
+	}
+
+	// Avoid the GCR error that complains if you try to delete an image which is
+	// referenced by a DockerManifestList, by first deleting all such manifest
+	// lists.
+	deleteManifestLists := deleteRequstsPopulator(isEqualTo(cr.DockerManifestList))
```

Comment:
It took me a bit to figure out what this line was doing in comparison to the `isEqualTo` anonymous function. Can you simplify `isEqualTo` to take in 2 parameters a & b, compare a& b, then return a bool instead of the nested functions.

Reply:
This is not possible AFAICS because we only know `a`, not `b`, at the time we make the call to `deleteRequestsPopulator`. That's why we're passing in a function (I guess you could say it's almost like a callback?) that can deal with `b` when we get it later down the line.